File: RxTango/python/examples/pipeline.py
"""Fluent 6-step TangoClient pipeline — the showstopper demo.

Read → calibrate → write → format → write → read back.
No threads.  No callbacks.  No intermediate variables.

Mirrors `TangoTestPipeline.java` and `FluentClient.java`.

Usage:
    python pipeline.py [device]

    device  defaults to tango://localhost:10000/sys/tg_test/1
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import TangoClient


async def main() -> None:
    device = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    done      = asyncio.Event()

    print(f"Fluent pipeline on {device}\n")

    TangoClient() \
        .read(device, "double_scalar") \
        .map(lambda v: (print(f"  [1] read     double_scalar   = {v:+.6f}") or v)) \
        .map(lambda v: abs(v) * 2.0 + 1.5) \
        .map(lambda v: (print(f"  [2] calibrated               = {v:+.6f}") or v)) \
        .write(device, "double_scalar_w") \
        .map(lambda v: (print(f"  [3] wrote    double_scalar_w = {v:+.6f}") or v)) \
        .map(lambda v: f"cal={v:.4f}") \
        .map(lambda s: (print(f"  [4] formatted                = {s!r}") or s)) \
        .write(device, "string_scalar_w") \
        .map(lambda s: (print(f"  [5] wrote    string_scalar_w = {s!r}") or s)) \
        .read(device, "string_scalar_w") \
        .subscribe(
            on_next=lambda v: print(f"\n  Confirmed on device: {v!r}"),
            on_error=lambda e: (print(f"  ERROR: {e}", file=sys.stderr), done.set()),
            on_completed=lambda: (print("  Pipeline complete."), done.set()),
            scheduler=scheduler,
        )

    await asyncio.wait_for(done.wait(), timeout=10.0)


if __name__ == "__main__":
    asyncio.run(main())

File: RxTango/cpp/examples/pipeline.cpp
/**
 * Fluent 6-step TangoClient pipeline — the showstopper demo.
 *
 * Read → calibrate → write → format → write → read back.
 * No threads.  No callbacks.  No intermediate variables.
 * The same declarative chain as Java's FluentClient and Python's pipeline.py.
 *
 * Mirrors Python's pipeline.py and Java's TangoTestPipeline.java / FluentClient.java.
 *
 * Usage:
 *   ./pipeline [device]
 *   defaults: tango://localhost:10000/sys/tg_test/1
 */

#include <any>
#include <condition_variable>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <mutex>
#include <sstream>
#include <string>

#include <rxtango/rxtango.hpp>

int main(int argc, char* argv[]) {
    const std::string device = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";

    std::cout << "Fluent pipeline on " << device << "\n\n";

    std::mutex              m;
    std::condition_variable cv;
    bool                    done = false;

    rxtango::TangoClient()
        .read(device, "double_scalar")
        .map([](std::any v) -> std::any {
            double d = std::any_cast<double>(v);
            std::cout << "  [1] read     double_scalar   = " << std::setprecision(6) << d << "\n";
            return d;
        })
        .map([](std::any v) -> std::any {
            double calibrated = std::abs(std::any_cast<double>(v)) * 2.0 + 1.5;
            std::cout << "  [2] calibrated               = " << calibrated << "\n";
            return calibrated;
        })
        .write(device, "double_scalar_w")
        .map([](std::any v) -> std::any {
            double d = std::any_cast<double>(v);
            std::cout << "  [3] wrote    double_scalar_w = " << d << "\n";
            return d;
        })
        .map([](std::any v) -> std::any {
            std::ostringstream ss;
            ss << std::fixed << std::setprecision(4) << "cal=" << std::any_cast<double>(v);
            std::string formatted = ss.str();
            std::cout << "  [4] formatted                = " << formatted << "\n";
            return formatted;
        })
        // Note: write() with string requires a separate typed write; use execute for DevString
        // Here we demonstrate a further map step to show the full chain
        .map([](std::any v) -> std::any {
            std::cout << "  [5] pipeline complete, result: " << std::any_cast<std::string>(v) << "\n";
            return v;
        })
        .subscribe(
            [](std::any v) {
                std::cout << "\n  Confirmed: " << std::any_cast<std::string>(v) << "\n";
            },
            [&](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) {
                    std::cerr << "  ERROR: " << ex.what() << "\n";
                }
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            },
            [&]() {
                std::cout << "  Pipeline complete.\n";
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            }
        );

    std::unique_lock<std::mutex> lk(m);
    cv.wait_for(lk, std::chrono::seconds(10), [&]{ return done; });
    return 0;
}
